src/lib/Serialize.hs

```diff
@@ -104,10 +117,81 @@ prettyVal val = case val of
     return $ align $ group innerDoc
   atom -> return $ prettyPrec atom LowestPrec
 
+-- === taking memory snapshots for serializing heap-backed dex values ===
+
+data WithSnapshot a = WithSnapshot a [PtrSnapshot]  deriving Generic
+type RawPtr = Ptr ()
+-- TODO: we could consider using some mmap-able instead of ByteString
+data PtrSnapshot = ByteArray BS.ByteString
+                 | PtrArray [PtrSnapshot]
+                 | NullPtr  deriving Generic
+
+class HasPtrs a where
+  traversePtrs :: Applicative f => (PtrType -> RawPtr -> f RawPtr) -> a -> f a
+
+takeSnapshot :: HasPtrs a => a -> IO (WithSnapshot a)
+takeSnapshot x =
+  liftM (WithSnapshot x) $ execWriterT $ flip traversePtrs x \ptrTy ptrVal -> do
+    snapshot <- lift $ takePtrSnapshot ptrTy ptrVal
+    tell [snapshot]
+    return ptrVal
+
+restoreSnapshot :: HasPtrs a => WithSnapshot a -> IO a
+restoreSnapshot (WithSnapshot x snapshots) =
+  flip evalStateT snapshots $ flip traversePtrs x \_ _ -> do
+    (s:ss) <- get
+    put ss
+    lift $ restorePtrSnapshot s
+
+takePtrSnapshot :: PtrType -> RawPtr -> IO PtrSnapshot
+takePtrSnapshot _ ptrVal | ptrVal == nullPtr = return NullPtr
+takePtrSnapshot (_, ptrTy) ptrVal = case ptrTy of
+  PtrType eltTy -> do
+    childPtrs <- loadPtrPtrs ptrVal
+    PtrArray <$> mapM (takePtrSnapshot eltTy) childPtrs
+  _ -> ByteArray <$> loadPtrBytes ptrVal
+
+restorePtrSnapshot :: PtrSnapshot -> IO RawPtr
+restorePtrSnapshot snapshot = case snapshot of
+  PtrArray  children -> storePtrPtrs =<< mapM restorePtrSnapshot children
+  ByteArray bytes    -> storePtrBytes bytes
+  NullPtr            -> return nullPtr
+
+loadPtrBytes :: RawPtr -> IO BS.ByteString
+loadPtrBytes ptr = do
+  numBytes <- fromIntegral <$> dexAllocSize ptr
+  liftM BS.pack $ peekArray numBytes $ castPtr ptr
+
+storePtrBytes :: BS.ByteString -> IO RawPtr
+storePtrBytes xs = do
+  let xs' = BS.unpack xs
+  ptr <- dexMalloc $ fromIntegral $ length xs'
+  pokeArray (castPtr ptr) xs'
```

Comment:
Going through list in here is pretty expensive (one heap allocation per copied byte!). Consider using [`unsafeUseAsCString`](https://hackage.haskell.org/package/bytestring-0.11.0.0/docs/Data-ByteString-Unsafe.html#v:unsafeUseAsCString) (which is safe in this case) and `memcpy`